performant-software/dm-2
# User authentication and authorization

### What does this PR do?
- Adds user account registration and authentication using devise-token-auth and redux-token-auth.
- Protects and scopes API endpoints according to token-based authentication credentials.
- Adds a site admin flag for users and a site admin pane visible to these users, enabling the approval and deletion of other user accounts.
- Adds project settings, including permissions table, publishing, and deletion.

### What issues does it address?
- Closes #1, closes #5, closes #6, closes #7, closes #27, closes #40 

### How to test
- At the [PR review app](https://dm-2-staging-pr-39.herokuapp.com/), log in to a site admin account (will send credentials via Slack) using the user menu and Sign In dialog from the upper right corner.
- Sign out of the admin account and create a new user account with a different email address and name, using the user menu -> Register dialog.
- On clicking Register, you should see a message that your account has been created but is pending approval by an admin. You should receive an email to your admin account notifying you that a new user has been created (may take a couple minutes).
- While signed in as the new user, you should see "(Pending Approval)" after the user's name in the upper right. The app should otherwise behave as if you were not signed in (only listing public projects, no Create Project/Document or document-editing buttons).
- Sign out of the new user account and sign back in to the site admin account. Pull up the Admin dialog from the user menu. You should see a table of all user accounts in the site, with the new, pending-approval account shaded pink at the top of the list. Clicking its toggle switch for Approved should send it into the main part of the list.
- Sign out of the admin account and back into the new user account. You should no longer see the pending approval suffix, and you should have the ability to create a new project.
- After clicking Create Project, you should see a gear symbol next to "Untitled Project" that will pull up a tabbed Settings dialog. In the first section, you should be able to update the project title and description — the former should be reflected after a pause in the Project nav bar, the latter should be reflected when you see the project in the home screen list.
- In the second tab, you should see a table of user permissions for the project. This should just include your current account, set to Admin for the project. You should not be able to change or revoke your own admin permission over the project. Clicking in the text field at the bottom of the table should pull up a keyboard-filterable list of all remaining users in the instance; selecting one, choosing a permission level, and clicking Add should add a new permission record to the table for that user.
- Accounts that have site admin status should be able to perform any action on any project, even if they are not listed in the project's permissions table.
- Accounts that have been given Read permission on a (non-public) project should see that project included in their home view list and be able to open it, but should not see Create Document buttons, document editing tools, or drag and drop targets.
- Accounts that have been given Write permission on a project should have full document creation and editing abilities, but should not see the gear icon to change project settings.
- Accounts that have been given Admin permission on a project should be able to perform any action within the project, including its settings dialog.
- The last tab in the project settings dialog contains a toggle to make the project public, as well as a checkbox-guarded button to delete the project. A public project should appear in the home view list and be viewable in read mode for all users and when signed out.
